{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8ee77f2-aef6-435a-8946-6e46ad8f6a70",
   "metadata": {},
   "source": [
    "First, implement a version of the naive exact matching algorithm that is strand-aware. That is, instead of looking only for occurrences of P in T, additionally look for occurrences of thereverse complement of P in T. If P is ACT, your function should find occurrences of both ACTand its reverse complement AGT in T.\n",
    "\n",
    "If P and its reverse complement are identical (e.g. AACGTT), then a given match offset should be reported only once. So if your new function is called naive_with_rc, then the old naivefunction and your new naive_with_rc function should return the same results when P equals its reverse complement.\n",
    "\n",
    "Next, download and parse the lambda virus genome, <a href=\"https://d28rh4a8wq0iu5.cloudfront.net/ads1/data/lambda_virus.fa\">here</a>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4be25855-b777-444b-bbec-e61a1f96a627",
   "metadata": {},
   "outputs": [],
   "source": [
    "#234567890#234567890#234567890#234567890#234567890#234567890#234567890#234567890\n",
    "DATA = '../data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5bf039f-81ad-40fa-8483-e3fc5bcfb44c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):  # loop over characters\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aece514b-549f-4db9-afab-2d126045b04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9561f89a-571a-4377-a869-42a3724fae0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def naive_with_rc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93487fee-0116-46fc-b83f-69f989a8a4c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ex 1\n",
    "p = 'CCC'\n",
    "ten_as = 'AAAAAAAAAA'\n",
    "t = ten_as + 'CCC' + ten_as + 'GGG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)   # 10, 23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0223413a-7278-4be8-b749-cb197909f7c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = 'CGCG'\n",
    "t = ten_as + 'CGCG' + ten_as + 'CGCG' + ten_as\n",
    "occurrences = naive_with_rc(p, t)\n",
    "print(occurrences)  # 10, 24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9a8801b-1ede-4a2e-b758-c3d9bdfd2e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genome(path):\n",
    "    genome = ''\n",
    "    with open(path, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ee0dcf0e-ceea-4573-aa1d-e2a1111bfd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "phix_genome = read_genome(f'{DATA}/phix.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302e673a-0a20-4f2c-8a0c-30251958be08",
   "metadata": {},
   "outputs": [],
   "source": [
    "occurrences = naive_with_rc('ATTA', phix_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efabbdec-6198-45da-8a25-512823537f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('offset of leftmost occurrence:' min(occurrences))  # 62\n",
    "print('N occurrences:', len(occurrences))  # 60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7860ed3-1df9-47b4-ab93-9318e960e0e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
